### PR TITLE
Adding .cadl & package.json cadlMain back compat support

### PR DESCRIPTION
**Pull request by allenjzhang**
- [x] Need to double check vscode .cadl file support.

### CODE REVIEW

File: packages/compiler/core/program.ts [hunk 1]
```diff
@@ -56,6 +56,12 @@ import {
   mutate,
 } from "./util.js";
 
+export function BackCompatSupportForCadlMain(packageJson: any) {
+  if (packageJson?.tspMain === undefined && packageJson?.cadlMain !== undefined) {
+    packageJson.tspMain = packageJson.cadlMain;
+  }
+}
```
Comment (nguerrera) [inline packages/compiler/core/program.ts:63]: I find it surprising that this mutates. 

I would prefer something like `export function getTspMain(packageJson: any) { return packageJson?.tspMain ?? packageJson?.cadlMain }`

Then replace `pkg?.tspMain` with getTspMain(pkg) throughout.

Also, I would move this to a util file where we don't export it publicly to everyone.

File: packages/compiler/core/program.ts [hunk 2]
```diff
@@ -899,7 +906,19 @@ export async function compile(
     const [pkg] = await loadFile(host, pkgJsonPath, JSON.parse, program.reportDiagnostic, {
       allowFileNotFound: true,
     });
-    const mainFile = resolvePath(dir, typeof pkg?.tspMain === "string" ? pkg.tspMain : "main.tsp");
+    // Back Compat: if main.cadl exist, return main.cadl
+    let mainFile = resolvePath(dir, "main.cadl");
```
Comment (nguerrera) [inline packages/compiler/core/program.ts:910]: It seems inconsistent to give main.cadl priority here.

I think the logic needs to be something like this:

```
// 1. Use tspMain (with fallback to cadlMain) 
const tspMain = getTspMain(pkg);
if (tspMain) {
 return tspMain;
}

// 2. Use main.cadl if it exists
/* ...(Put this back compat code here)... */

// 3. Use main.tsp
return "main.tsp";

File: packages/typespec-vs/src/VSExtension.cs [hunk 3]
```diff
@@ -148,23 +152,21 @@ private void LogStderrMessage(string? message)
         }
 
 #if DEBUG
-        private static bool InDevelopmentMode()
-        {
-            return string.Equals(
-              Environment.GetEnvironmentVariable("TYPESPEC_DEVELOPMENT_MODE"),
-              "true",
-              StringComparison.OrdinalIgnoreCase);
+        private static bool InDevelopmentMode () {
```
Comment (nguerrera) [inline packages/typespec-vs/src/VSExtension.cs:155]: Some unexpected reformatting in this file, please revert. I think automated format in PR validation is not handling the `#if DEBUG`, but I would expect VS to use our .editorconfig, so not sure how this got reformatted for you.

Comment (allenjzhang) [reply packages/typespec-vs/src/VSExtension.cs]: Updated with VS2022 format. Still resulted in a few changes. But those changes look good.

File: packages/typespec-vs/src/VSExtension.cs [hunk 4]
```diff
@@ -39,6 +39,11 @@ public sealed class ContentDefinition
         [FileExtension(".tsp")]
         [ContentType("typespec")]
         public FileExtensionToContentTypeDefinition? TypeSpecFileExtensionDefinition => null;
+
+        [Export]
+        [FileExtension(".cadl")]
+        [ContentType("typespec")]
+        public FileExtensionToContentTypeDefinition? CadlFileExtensionDefinition => null;
```
Comment (nguerrera) [inline packages/typespec-vs/src/VSExtension.cs:46]: It looks like the changes to FilesToWatch got reverted accidentally.

**Review by nguerrera — APPROVED** [review]
Just the one comment about FilesToWatch not having *.cadl and cadl-project anymore after reformatting.

**Comment by nguerrera** [comment]
Oh, it auto-merged :)

**Comment by nguerrera** [comment]
Opened https://github.com/microsoft/typespec/pull/1675 to fix the one issue.

**Pull request merged**